simphony_mayavi/plugins/add_source_panel.py

```diff
@@ -0,0 +1,177 @@
+        ''' Text for representing the EngineSource '''
+        # All the point/cell data names being added
+        source = getattr(object, trait)[index]
+        data_names = ",".join([name for name in (source.point_scalars_name,
+                                                 source.point_vectors_name,
+                                                 source.cell_scalars_name,
+                                                 source.cell_vectors_name)
+                               if name])
+        text = "{source}({data}) from {engine}"
+        return text.format(source=source.dataset,
+                           data=data_names,
+                           engine=source.engine_name)
+
+
+class PendingEngineSourceHandler(Handler):
+    ''' UI Handler for adding dataset as a source in EngineSourceManager
+    The source is appended to the ``_pending_engine_sources`` list '''
+
+    def append_list(self, info):
+        ''' Confirm and add the EngineSource to the pending list '''
+        info.manager._pending_engine_sources.append(info.object)
+        info.ui.dispose()
+
+
+class AddSourcePanel(HasTraits):
+    ''' Standalone UI for adding datasets from a modeling engine to
+    a Mayavi scene
+
+    Parameters
+    ----------
+    engine : Instance of ABCModelingEngine
+        Simphony Modeling Engine wrapper
+    engine_name : str
+        Name of the modeling engine
+    mayavi_engine : Instance of mayavi.core.engine.Engine
+        for visualization
+    '''
+    engine = Instance(ABCModelingEngine)
+    engine_name = Str
+    mayavi_engine = Instance("mayavi.core.engine.Engine")
+
+    label = "Add to Mayavi"
+
+    # Buttons for the UI
+    _add_dataset = Button("+")
+    _remove_dataset = Button("-")
+    _add_to_scene = Button("Send to Scene")
+
+    # Pending EngineSource to be added to Mayavi
+    _pending_engine_sources = List(Instance(EngineSource))
+
+    # Selected pending EngineSource (as an index of the list)
+    _pending_source = Enum(values="_pending_engine_sources")
+    _pending_source_index = Int
+
+    panel_view = View(
+        VGroup(
+            HGroup(
+                Item("_add_dataset", show_label=False),
+                Item("_remove_dataset", show_label=False,
+                     enabled_when="_pending_source"),
+                Item("_add_to_scene", show_label=False,
+                     enabled_when="_pending_engine_sources")),
+            Item("_pending_engine_sources",
+                 show_label=False,
+                 editor=ListStrEditor(
+                     adapter=EngineSourceAdapter(),
+                     editable=False,
+                     selected="_pending_source",
+                     selected_index="_pending_source_index")),
+            ),
+        title="Visualize")
+
+    # --------------------------------------------------
+    # Public methods
+    # --------------------------------------------------
+
+    def __init__(self, engine_name, engine, mayavi_engine):
```

Comment:
you typically avoid doing this with `HasTraits` class. In this case, you break the traits initialization and event system. 

You can remove the `__init__` method and use the standard trait intialization by keyword:

```
AddSourcePanel(engine=engine, engine_name=engine_name, mayavi_engine=mayavi_engine)
```

If you really, really need an `__init__`  method, you should calls `HasTraits.__init__(self, **kwargs)` to properly intialize the `HasTraits` class (with the appropriate definitions of the `kwargs`)